core/src/main/java/org/infinispan/persistence/util/PersistenceManagerCloseableSupplier.java

```diff
@@ -124,6 +124,11 @@ public void processEntry(MarshalledEntry<K, V> marshalledEntry, AdvancedCacheLoa
          closeLock.lock();
          try {
             if (closed) {
+               // If is possible that this supplier was closed right after we polled the queue. In that case they may
+               // have also inserted a value. We need to double check that the queue is REALLY empty after being closed.
+               if (queue.peek() != null) {
```

Comment:
could you just do a poll and, if the entry is not null, return it (instead of continue?)